numpy/random/mtrand/mtrand.pyx

```diff
@@ -4899,10 +4899,20 @@ cdef class RandomState:
                [3, 4, 5]])
 
         """
-        if isinstance(x, (int, long, np.integer)):
-            arr = np.arange(x)
-        else:
-            arr = np.array(x)
+        if not isinstance(x, (int, long, np.integer)):
+            arr = np.asarray(x)
+            
+            # shuffle has fast-path for 1-d
+            if x.ndim == 1:
+                # must return a copy
+                if arr is x:
+                    arr = np.array(arr)
+                self.shuffle(arr)
+                return arr
+
+            return arr[self.permutation(x.shape[0])]
```

Comment:
Why not just use shuffle here directly instead of a recursive call?

Reply:
That way you can also make the `arange` type specific to this case.

Reply:
I thought the recursion was tidy, but it probably isn't